runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.telemetry.otel
+
+import aws.smithy.kotlin.runtime.telemetry.context.Context
+import aws.smithy.kotlin.runtime.telemetry.metrics.*
+import aws.smithy.kotlin.runtime.util.Attributes
+import io.opentelemetry.api.OpenTelemetry
+import io.opentelemetry.api.metrics.ObservableDoubleMeasurement
+import io.opentelemetry.api.metrics.ObservableLongMeasurement
+import io.opentelemetry.api.metrics.DoubleHistogram as OtelDoubleHistogram
+import io.opentelemetry.api.metrics.LongCounter as OtelLongCounter
+import io.opentelemetry.api.metrics.LongHistogram as OtelLongHistogram
+import io.opentelemetry.api.metrics.LongUpDownCounter as OtelLongUpDownCounter
+import io.opentelemetry.api.metrics.Meter as OpenTelemetryMeter
+
+internal class OtelMeterProvider(private val otel: OpenTelemetry) : MeterProvider {
+    override fun getOrCreateMeter(scope: String, attributes: Attributes): Meter {
+        val meter = otel.getMeter(scope)
+        // FIXME: meter level attributes not supported yet by Java impl
+        // https://github.com/open-telemetry/opentelemetry-java/issues/5503
+        return OtelMeter(meter)
+    }
+}
+
+private class OtelMeter(
+    private val otelMeter: OpenTelemetryMeter,
+) : Meter {
+    override fun createUpDownCounter(name: String, units: String?, description: String?): UpDownCounter {
+        val counter = otelMeter.upDownCounterBuilder(name)
+            .apply {
+                description?.let { setDescription(it) }
+                units?.let { setUnit(units) }
+            }
+            .build()
+        return OtelUpDownCounterImpl(counter)
```

Comment:
**Question**: Are the `?.let` null-checks necessary? The property values on new builders should already be `null` so re-setting them to `null` shouldn't have any negative effect:

```kotlin
val counter = otelMeter
    .upDownCounterBuilder(name)
    .setDescription(description)
    .setUnit(units)
    .build()
return OtelUpDownCounterImpl(counter)
```

Reply:
They are:

```
w: file:///smithy-kotlin/runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:38:29 Type mismatch: inferred type is String? but String was expected
```

Reply:
well at least if you want to get past warnings as errors check